[team-15] oauth 구현

choigiseong opened the pull request:
안녕하세요. 백엔드 반의 coco입니다.
리뷰를 원하는 특정 코드만 코멘트 달아 놓을 테니 중점적으로 봐주셨으면 감사하겠습니다.

* GetPropetyUtil  은 applicationContext에서 property를 얻어오는 유틸입니다.
* GitHubDeviceType enum은 ios와 fe의 gitOauthApp을 분리 시키기 위한 enum입니다.
* Oauth는 외부 라이브러리로 만들고 주입했습니다. GithubOauth는 Oauth를 상속해서 사용합니다
* https://github.com/ChoiGiSung/gitOauth/tree/main/gitOauthV1/src/main/java
    라이브러리로 사용한 Oauth입니다

부족한 코드지만 봐주셔서 감사합니다.

choigiseong committed:
feat : init

choigiseong committed:
feat : 초기 세팅

choigiseong committed:
feat : docker 세팅

choigiseong committed:
feat : git ignore 수정

- application.properties 제외

choigiseong committed:
feat : git ignore 적용

choigiseong committed:
feat : db 스키마 변경, localhost 변경

- 문법 오류 때문에 변경

choigiseong committed:
feat : wish_list table 추가

choigiseong committed:
feat : 숙소의 이미지를 저장할 table

choigiseong committed:
feat : 예약을 위한 class 정의

choigiseong committed:
feat : 도시정보를 저장할 class 정의

choigiseong committed:
feat : 호스의 정보를 저장할 class 정의

choigiseong committed:
feat : rooms 도메인에 관련된 class들 정의

choigiseong committed:
feat : user 도메인에 관련된 class들 정의

choigiseong committed:
feat : localDate로 변경

- localDateTime에서 localDate로 변경

choigiseong committed:
feat : SearchInfoDTO 정의

- price를 위한 검색DTO정의

choigiseong committed:
feat : user table명 명시

- users로 변경

choigiseong committed:
feat : 검색 사항을 저장할 DTO 정의

- DTO를 정의
- 검색 조건이 없으면 기본값으로 세팅

choigiseong committed:
feat : VO로 쓸 class 선언 money와 guest

- VO로 사용할 예정이다.

choigiseong committed:
feat : VO에 final 추가

- VO이니 final 추가

choigiseong committed:
feat : pricePerDate의 타입 변경

- money로 변경

choigiseong committed:
feat : controller 정의

- 기본 기능 정의

choigiseong committed:
feat : DTO간 상속관계 설정

choigiseong committed:
feat : city table의 기본키 변경

choigiseong committed:
feat : 코틀린 sql 작성

choigiseong committed:
feat : 검색조건으로 prices 검색

- 조건에 맞는 가격들을 검색한다

choigiseong committed:
feat : 예시 data 추가

choigiseong committed:
feat : getter와 생성자, builder 생성

- mapper에서 사용할 builder와 생성자를 만들었다

choigiseong committed:
feat : rooms DB에서 읽어오기

- searchRoomsDTO를 통해 해당하는 객체를 읽어옵니다

choigiseong committed:
feat : wish 추가

- rooms는 유저의 wish를 저장한다

choigiseong committed:
feat : search시 SearchRoomsListDTO를 반환

- 모두 dto로 변환하고 반환합니다

choigiseong committed:
feat : roomsDetail을 얻기

- rooms id에 해당하는 방을 얻는다
- roomsDetail을 정의하고 안에 있는 것을 다 DTO로 변환한다

choigiseong committed:
feat : 예시 데이터 올리기

- 외부 api를 java로 가공해서 데이터 만들었음

choigiseong committed:
feat : 금액을 decimal에서 int로 변경

- decimal이 맞지만 오류가 많이 나서 일단 변경

choigiseong committed:
refactor : 책임 분산

- controller에서 DTO로 변환하는 것이 아닌 service에서 변환
- 기타 이름들 변경

choigiseong committed:
fix : 잘못된 sql 고치기

- 검색 조건이 오작동 하고 있었다.

choigiseong committed:
refactor : rooms를 room으로 이름 변경

- 모든 곳에서 이름을 변경했습니다
- rooms는 방들을 의미하게 되었습니다

choigiseong committed:
refactor : 메소드 이름 변경

- 일관성 있게 변경

choigiseong committed:
refactor : Search시 생성자에서 null 체크 리팩토링

- 팩토리 메소드로 이름을 부여했다
- String을 LocalDate로 변화하는 로직은 util로 빼서 사용한다

choigiseong committed:
refactor : 상수로 추출

- 특정 값을 상수로 추출
- 이름이 더 알맞게 수정

choigiseong committed:
chore : 패키지 이동

- sql 패키지로 이동

choigiseong committed:
feat : 모든 도시들 보여주기

- 모든 도시를 보여줍니다

choigiseong committed:
refactor : 메소드 이름 변경

- 더 의미 있게 이름 변경

choigiseong committed:
fix : 오타 수정

choigiseong committed:
feat : 예약하기 기능 구현

- 아직 부족한 검증이 많지만 기본틀을 작성

choigiseong committed:
feat : data 변경

choigiseong committed:
refactor : 여러 이름 변경

- url을 더 rest스럽게 변경
- 복수의 값을 전달하면 복수형으로 이름을 정한다

choigiseong committed:
refactor : RoomDAO 정리

- 이미지와 예약을 채워주는 함수를 작성
- optional 함수 삭제

choigiseong committed:
feat : 방 상세보기 시 예약된 내역을 보여줘야 한다

- 그래서 예약되어있는 날짜들을 보여준다

choigiseong committed:
feat : totalPrice를 계산해서 준다

- RoomPreview에서는 총 숙박비를 받아야 합니다

choigiseong committed:
chore : 패키지 이동

- sql 패키지 이동

choigiseong committed:
Merge pull request #18 from ChoiGiSung/feature_BE/12

Feature be/12

choigiseong committed:
feat : 예약시 예약을 검증한다

- 예약 검증을 위해 도메인과 서비스에서 각종 체크를 한다
- 아직 oauth를 구현하지 않아서 user는 null로 사용한다

choigiseong committed:
feat : CalcUtil 구현

- 계산을 하는 유틸 구현

choigiseong committed:
feat : 비즈니스 예외 정의

- 예약 시 수용인원이 초과되었을 때, 돈이 안맞을 때, 예약날짜가 이미 있을 때

choigiseong committed:
fix : 계산 오류 수정

- 더블로 하지 않아서 백분율계산이 되지 않았다

choigiseong committed:
feat : exception Handler와 ErrorReason을 구현

- 예외를 컨트롤하는 곳과 예외이유를 보여주는 class 정의

choigiseong committed:
feat : 찾는 방이 없을 경우 예외

- 없는 방을 찾는 경우 예외

choigiseong committed:
style : 빈 줄 삭제

choigiseong committed:
chore : 패키지 이동

- 패키지 이동

choigiseong committed:
feat : 예약 삭제 구현

- 소프트 딜리트로 구현, 그래서 기존 예약시 상태를 넘기기로 함
- 방 조회시에도 예약이 캔슬된 애들도 보여줌
- userId와 roomId,reservationId로 예약을 판단함

choigiseong committed:
feat : 데이터 추가

choigiseong committed:
feat : 기본 생성자 추가

choigiseong committed:
Merge pull request #20 from ChoiGiSung/feature_BE/19

Feature be/19

choigiseong committed:
feat : 불필요 getter 삭제

choigiseong committed:
feat : 다른 예약의 체크아웃 날짜에 체크인이 가능하게 변경

- 날짜를 조정해서 체크아웃하는 날에도 체크인을 가능하게했다
- 근데 java코드에서는 이미 가능했고 조회로직만 변경한 것이다

choigiseong committed:
feat : validation 에러 잡기 및 에러 코드 재정의

- 각각에 맞는 http status code를 정했습니다

choigiseong committed:
feat : default 날짜를 짧게 설정

choigiseong committed:
test : 예약 요구 조건 테스트

- 여러 요구조건을 test

choigiseong committed:
refactor : 추가비용 책임을 옮기기

- 책임은 additionalCost가 가지는게 맞을 거 같다

choigiseong committed:
refactor : totalPrice 리팩토링

- 책임을 좀 더 분산했습니다

choigiseong committed:
refactor : 수용인원 책임 분리

- 계산 책임을 option에게 주었다

choigiseong committed:
refactor : 메소드 이름 변경

- 수용인원 체크가 더 적절해 보인다

choigiseong committed:
Update README.md

choigiseong committed:
feat : Oauth를 상속받는 GitOauth 구현

- 라이브러리로 만든 Oauth를 상속 받아서 사용한다

choigiseong committed:
feat : GitOauth를 이용한 로그인 구현

- gitOauth를 사용해서 로그인 한다

choigiseong committed:
feat : serverKey 제외

choigiseong committed:
chore : 패키지 이동

choigiseong committed:
feat  : 인터셉터 구현으로 특정 uri만 로그인 검증

- 예약,예약 취소 시 검증
- 예외도 만들었다

choigiseong committed:
feat : HandlerMethodResolver 등록

- jwt에서 userId를 얻기 위해서 resolver를 생성

choigiseong committed:
Merge branch 'team-15' of https://github.com/ChoiGiSung/airbnb into feature_BE/24

choigiseong committed:
feat : 상태 코드 변경

choigiseong committed:
fix : requestParm 추가

choigiseong committed:
feat : api 버저닝

- 버전을 추가했다

choigiseong committed:
feat : 로그인 시 converter 등록

- uri상에서 구분하기 위해 ios와 desktop을 구분
- 둘을 enum으로 받기 위해 converter등록

choigiseong committed:
refactor : 예외처리 수정

- error code 를 만들어 메시지들을 관리
- 계층 구조를 만들어서 예외를 관리

choigiseong committed:
refactor : 예약 검증 책임 분리

- 도메인 로직으로 구현

choigiseong committed:
feat : jwt 예외 핸들 추가

- jwt의 예외를 처리한다

choigiseong committed:
feat : 예외 컨트롤, cors 에러 제어

choigiseong committed:
feat : enum 설정시 값을 변경하기

- enum 초기화시에 property에서 값을 꺼내 세팅한다

choigiseong committed:
Merge pull request #35 from ChoiGiSung/feature_BE/24

[oauth] oauth 기능 구현 끝

choigiseong commented on backend/src/main/java/com/codesquad/coco/oauth/gitoauth/GitHubDeviceType.java:
```diff
@@ -0,0 +1,44 @@
+
+public enum GitHubDeviceType {
+
+    IOS("github.ios.client.id", "github.ios.secret", "http://localhost:8080/git/auth/app"),
+    DESKTOP("github.desktop.client.id", "github.desktop.secret", "http://localhost:8080/git/auth");
+
+    private String clientKey;
+    private String clientSecret;
+    private String redirectUri;
+
+    GitHubDeviceType(String clientKey, String clientSecret, String redirectUri) {
+        this.clientKey = GetPropertyUtil.getProperty(clientKey);
+        this.clientSecret = GetPropertyUtil.getProperty(clientSecret);
+        this.redirectUri = redirectUri;
+    }
```
enum 초기화 부분입니다. util을 통해 secretKey.properties에서 값을 꺼내 초기화 시키는 로직입니다.
위의 코드로 인해 enum은 비밀 키를 가질 수 있게 되었습니다.

choigiseong committed:
fix : change type 변경

- 타입이 변경 되면 client id 가 변경된다

choigiseong committed:
Merge branch 'feature_BE/24' into backend

choigiseong commented on backend/src/main/java/com/codesquad/coco/oauth/gitoauth/GitOauth.java:
```diff
@@ -0,0 +1,39 @@
+        return getTemplate().exchange(userInfoUri, HttpMethod.GET, userInfo, GitUserInfoDTO.class).getBody();
+    }
+
+    public void changeType(GitHubDeviceType gitHubDeviceType) {
+        this.clientId = gitHubDeviceType.getClientKey();
+        this.clientSecret = gitHubDeviceType.getClientSecret();
+        this.redirectUri = gitHubDeviceType.getRedirectUri();
```
converter로 받은 enum에서 비밀 키를 꺼내서 사용합니다.
enum은 생성될 때 property를 가지고 와서 초기화 합니다

choigiseong commented on backend/src/main/java/com/codesquad/coco/room/model/Room.java:
```diff
@@ -0,0 +1,265 @@
+    }
+
+
+    public boolean reservationAvailability(ReservationDTO reservationDTO) {
+        int fewNights = LocalDateUtil.getAccommodationDay(reservationDTO.getCheckIn(), reservationDTO.getCheckOut());
+        int totalPrice = calcTotalPrice(fewNights);
+
+        reservationDateCheck(reservationDTO.getCheckIn(), reservationDTO.getCheckOut());
+        capacityCheck(reservationDTO.getAdult(), reservationDTO.getChild());
+        if (totalPrice != reservationDTO.getTotalPrice()) {
+            throw new TotalPriceNonMatchException(totalPrice);
+        }
+        return true;
+    }
```
예약 검사 로직입니다. 
날짜와 인원 수 , 총 금액을 검사하는 로직 입니다.
 domain에서 모든 검사를 하는 메소드를 만들어 이것만 노출 시키는 것이 좋은지.
아님 service에서 각 검사 별로 따로 노출 시키는 게 좋을 지 고민이었습니다.

현재는 모든 검사를 하는 하나의 메소드 노출 시켜 사용하고 있습니다.

ksundong replied:
`int fewNights = LocalDateUtil.getAccommodationDay(reservationDTO.getCheckIn(), reservationDTO.getCheckOut());`
일단 이렇게 계산하는 부분 별로 맘에 들지 않구요.
getter 사용하지않고 계산해보세요.

ksundong replied:
예약 검사를 하는 로직이고, 말씀하신게 모두 필요하다면 당연히 이쪽에서 해주는게 맞습니다.
다만, 이런 식의 메서드 추출은 좋은 형태가 아니라고 보여지네요.

1. 날짜 검사는 Reservation List를 이용하면 되지 않을까요? 일급 컬렉션을 사용하면 좋을 것 같네요.
2. 인원 수에 대한 검사는 인원수를 담당하는 클래스가 수행하면 되겠죠.
3. 총 금액 역시 총 금액을 담당하는 클래스가 수행하면 되겠죠.

결국 객체지향적으로 설계한다면, 내가 처리할 수 없는 책임은 다른 메서드를 추출해서 해결하는 것이 아니라, 다른 객체에게 책임을 위임하는 식으로 설계하는 것이 맞습니다.

choigiseong replied:
DTO에 로직을 추가하면 안되는 줄 알았는데, 찾아보니 간단한 값 변환이나 유효성 검사는 괜찮다고 해서 변경했습니다.
DTO에 숙박일을 계산하는 메소드를 추가해 줬습니다.

---


검사하는 로직 부분입니다.
 `reservationAvailability`에서 사용하는 메소드 `capacityCheck`와 `reservationDateCheck`등은 실제로 밑에서 다음과 같이 동작했습니다.

>     public boolean capacityCheck(int adult, int child) {
>         if (!roomOption.capacityCheck(adult, child)) {
>             throw new OvercapacityException();
>         }
>         return true;
>     }
> 
>     public boolean reservationDateCheck(LocalDate checkIn, LocalDate checkOut) {
>         for (Reservation reservation : reservations) {
>             if (!reservation.reservationDateCheck(checkIn, checkOut)) {
>                 throw new NonReservationException();
>             }
>         }
>         return true;
>     }

체크 같은 로직은 해당 객체에 위임하고, 그 로직을 행하는 메소드를 `room`객체에 정의되어있었습니다.
변경된 점 :  예약일 검사는 일급컬렉션안으로, 인원 체크는 `roomOption`안으로 이동 시켜 그 안에서 예외를 터트리게 했습니다. 감사합니다

7a57904 , 9d164c9



금액 계산 로직은 할인비 용,추가비 용 따로 계산되어야 할 것 같아서 2개의 메소드로 분리하고 따로 호출했었습니다.
이 부분은 위의 조언에 따라 하나의 메소드만 호출하게 변경했습니다.

39d4cd4

choigiseong reviewed (commented):
리뷰를 원하는 로직들 입니다. 나머지는 crud 수준이라 밑의 코드에 대한 리뷰어님의 생각을 알고 싶습니다.

ksundong commented:
제가 원래 오늘 리뷰하려고 했는데 사정이 있어서 내일 하겠습니다! 죄송합니다!

ksundong commented on backend/build.gradle:
```diff
@@ -0,0 +1,38 @@
+    testImplementation 'org.springframework.boot:spring-boot-starter-test'
+    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk8"
+    implementation "org.springframework.boot:spring-boot-starter-validation"
+    implementation files('build/libs/gitOauthV1.jar')
```
직접 만들어 주신건가요?

choigiseong replied:
넵! 맞습니다. 저번 미션에 라이브러리로 만들어 보라는 옵션이 있어서 그 때 만들고 이번에 재사용을 해봤습니다.

ksundong commented on backend/docker/db/conf/custom.cnf:
```diff
@@ -0,0 +1,4 @@
+[mysqld]
+character-set-server = utf8
+collation-server = utf8_unicode_ci
```
해당 charset을 사용하는건 현 시점에선 추천드리지 않습니다.

ksundong commented on backend/src/main/java/com/codesquad/coco/city/CitiesDTO.java:
```diff
@@ -0,0 +1,16 @@
+
+public class CitiesDTO {
+
+    private List<CityDTO> cities;
```
Collection은 널 방지를 해주는 것을 추천합니다.

choigiseong replied:
수정했습니다.

ksundong commented on backend/src/main/java/com/codesquad/coco/city/CityController.java:
```diff
@@ -0,0 +1,22 @@
+@RestController
+public class CityController {
+
+    private CityService cityService;
```
```suggestion
    private final CityService cityService;
```

ksundong commented on backend/src/main/java/com/codesquad/coco/city/CityController.java:
```diff
@@ -0,0 +1,22 @@
+        this.cityService = cityService;
+    }
+
+    @GetMapping("/cities")
```
이정도는 미래지향적인 설계로 RequestMapping을 추출해보는것도 나쁘지않다고 생각합니다.

ksundong commented on backend/src/main/java/com/codesquad/coco/city/CityDAO.java:
```diff
@@ -0,0 +1,23 @@
+
+import static com.codesquad.coco.utils.sql.CitySQLKt.FIND_ALL;
+
+@Component
```
Repository 애노테이션을 보통 많이 사용하는데요
둘의 차이를 학습해보세요.

choigiseong replied:
Repository를 사용하지 않았던 이유는 DAO와 Repository는 같지 않다고 생각해서였습니다. 그래서 마땅한 게 없어 더 상위인 component를 사용했습니다.
 dao는 table 하나에 해당하고, repository는 루트 애그리거트를 위한 것이기 때문입니다.

근데 이번에 찾아보니 프리젠테이션 레이어에서는 Repository를 사용하라고 하네요.
 다음부터는 repository를 사용하도록 하겠습니다.

ksundong replied:
프리젠테이션 레이어가 아니라 퍼시스턴트 레이어일겁니다 ㅎㅎ 착각하신 것 같아요.
코코가 말씀하시는 Repository라는 용어는 DDD의 용어입니다.
거기의 영향을 받은게 Spring Data의 Repository들이죠.
Repository 애노테이션은 그 의미는 없는 것 같아서 DAO에 사용해주시면 될 것 같고, DAO 역시 table 하나에 해당되지 않는다고 생각해요. (만약 하나만 담당해야한다면 조인을 쓰면 안되겠죠?)

ksundong commented on backend/src/main/java/com/codesquad/coco/city/CityDAO.java:
```diff
@@ -0,0 +1,23 @@
+    }
+
+    public List<City> findAll(){
+        return template.query(FIND_ALL,new CityMapper());
```
```suggestion
        return template.query(FIND_ALL, new CityMapper());
```

ksundong commented on backend/src/main/java/com/codesquad/coco/city/CityDTO.java:
```diff
@@ -0,0 +1,20 @@
+package com.codesquad.coco.city;
```
패키징을 좀 더 잘 해보시면 좋을 것 같아요.
모든 것이 한 패키지에 있는건 별로 추천드리고 싶지 않은 형태네요~

choigiseong replied:
현재는 하나의 도메인 안에 관련된 것을 다 넣었는데, 
다음에는 계층 별로 패키징을 해보겠습니다.

ksundong commented on backend/src/main/java/com/codesquad/coco/city/CityService.java:
```diff
@@ -0,0 +1,25 @@
+    public List<CityDTO> findAllCityDTOs(){
+        List<City> cities = cityDAO.findAll();
+        return cities.stream()
+                .map(DTOConverter::cityToCityDTO)
```
컨버터가 꼭 필요할까요?

ksundong replied:
생성자 내지는, 정적 팩토리 메서드를 활용해보면 어떨까 싶네요~

choigiseong replied:
하나의 메소드를 통과해서 공통적으로 처리하고 싶었는데. 변화에 쉽게 대응하지 못하겠네요.
컨버터를 삭제하고 각 DTO에서 정적 팩토리 메소드를 정의했습니다.

ksundong commented on backend/src/main/java/com/codesquad/coco/global/ApplicationContextServe.java:
```diff
@@ -0,0 +1,28 @@
+import org.springframework.stereotype.Component;
+
+@Component
+public class ApplicationContextServe implements ApplicationContextAware {
```
왜 사용해주셨는지 궁금해지는군요.

choigiseong replied:
이 부분은 최종적으로  `GitHubDeviceType`을 위해서 작성했습니다.

해당 class를 통해 `applicationContext`를 얻고, `GetPropertyUtil`에서 컨텍스트를 활용해서 `properties`를 읽으려 했습니다.

`GitHubDeviceType`에서 활용하는 부분입니다.

>     IOS("github.ios.client.id", "github.ios.secret", "http://localhost:8080/git/auth/app"),
>     DESKTOP("github.desktop.client.id", "github.desktop.secret", "http://localhost:8080/git/auth");
> 
>     GitHubDeviceType(String clientKey, String clientSecret, String redirectUri) {
>         this.clientKey = GetPropertyUtil.getProperty(clientKey);
>         this.clientSecret = GetPropertyUtil.getProperty(clientSecret);
>         this.redirectUri = redirectUri;
>     }

ios와 desktop을 구분 지은 이유는 각각 다른 app이기 때문입니다. 그렇기에 다른 `clientId`가 있습니다.
이렇게 구현한 이유는 `redirectUri ` 때문이었습니다. git은 하나의 redirectUri만 등록할 수 있더라고요.
어떻게 해결할까 하다가 인프런이 위와 같이 다른 app을 사용하는 것을 보고  코드로 구현해봤습니다.

이렇게 되면 사용자는 로그인 시에 device 타입만 바꾸면 다른 app으로 접속이 가능합니다.
```
    @PostMapping("/v1/auth/{gitHubDeviceType}")
    public JWT login(@RequestParam String code, @PathVariable GitHubDeviceType gitHubDeviceType) {
        GitUserInfoDTO userInfo = userService.loginByGitOauth(code, gitHubDeviceType);
        String jwt = JWTUtils.createJWTTypeBearer(userInfo, serverKey.getJwtServerKey());
        return new JWT(jwt);
    }
```

제가 궁금했던 점은 enum에서 직접 작성한 값을 생성자에서 활용해 초기화하는 부분이었습니다.
이렇게 구현 해본 적이 없어서 괜찮은 방법인지 아니면 복잡도만 증가한 것인지 고민입니다.

ksundong replied:
애플리케이션 컨텍스트를 클래스 레벨에서 읽으려 하는 것 부터 위험하다고 생각되긴 하네요.
JVM의 동작원리와 Spring이 어떻게 동작하는지에 대해서 안다면 아마 다른 접근 방법을 취하지 않을까 싶어요.
사실 디바이스 타입을 직접 넘겨주는 것 자체가, 그렇게 좋은 설계인가? 에 대해서 생각해보시면 좋을 것 같아요.
왜 타입별로 구분이 필요할까요?

choigiseong replied:
타입 구분이 필요한 이유는 밑의 client Id가 왜 달라야 하는가 에서 설명하겠습니다.

ksundong commented on backend/src/main/java/com/codesquad/coco/global/auth/UserIdHandlerMethodResolver.java:
```diff
@@ -0,0 +1,32 @@
+    @Override
+    public Long resolveArgument(MethodParameter methodParameter, ModelAndViewContainer modelAndViewContainer, NativeWebRequest nativeWebRequest, WebDataBinderFactory webDataBinderFactory) throws Exception {
+        String jwt = nativeWebRequest.getHeader(JWTUtils.HEADER_TYPE);
+        Object infoFromJWT = JWTUtils.getInfoFromJWT(jwt, serverKey.getJwtServerKey(), WANT_INFO);
```
JWtUtils에 id를 얻어오는 함수를 만들어주는 것도 괜찮았겠네요.

ksundong commented on backend/src/main/java/com/codesquad/coco/global/exception/CustomExceptionHandler.java:
```diff
@@ -0,0 +1,92 @@
+    /**
+     * 비즈니스 예외
+     **/
+
```
여기 빈 줄은 왜 띄워주신걸까요?

choigiseong replied:
다 비슷한 주석이라 복붙하다 생긴 빈 줄 같네요. 수정했습니다!

ksundong commented on backend/src/main/java/com/codesquad/coco/global/exception/CustomExceptionHandler.java:
```diff
@@ -0,0 +1,92 @@
+        return new ResponseEntity<>(errorReason, HttpStatus.valueOf(errorCode.getHttpStatus()));
+    }
+
+    /**
+     * jwt 예외
+     **/
+    @ExceptionHandler(OverTimeException.class)
```
주석과 예외명이 매칭이 되지는 않는 것 같네요.

choigiseong replied:
맞습니다. jwt를 구현할 때 더 많은 예외를 잡았어야 했는데, 하나의 예외만 잡았습니다..
수정하게 된다면 jwt 예외를 상위로 두고 그것을 공통적으로 잡아 처리하겠습니다.

ksundong commented on backend/src/main/java/com/codesquad/coco/global/exception/CustomExceptionHandler.java:
```diff
@@ -0,0 +1,92 @@
+     * jwt 예외
+     **/
+    @ExceptionHandler(OverTimeException.class)
+    @ResponseStatus(HttpStatus.FORBIDDEN)
```
403을 기본으로 주는게 맞는지 고민해보세요.

ksundong replied:
403에러는 언제 반환되어야 할까요?

choigiseong replied:
401은 로그인되어 있지 않을 때
403은 권한이 없는 곳에 요청할 때라고 이해하고 있습니다.

overTime은 jwt가 만료된 것이기에 로그인이 안 돼 있다고 판단하고 401이 주는 게 좋을 거 같습니다.

복붙을 하다 보니 수정 안 된 곳이 많네요. 주의하겠습니다.

ksundong commented on backend/src/main/java/com/codesquad/coco/global/exception/ErrorCode.java:
```diff
@@ -0,0 +1,58 @@
+    /**
+     * 비즈니스 예외
+     **/
+    RESERVED_FOR_THAT_DATE(422, "그 날에는 예약이 이미 있음"),
+    OVER_CAPACITY(422, "수용 인원 초과"),
+    TOTAL_PRICE_NON_MATCH(422, "클라이언트의 총액과 일치하지 않음 : "),
```
상세한 에러코드의 사용 좋네요.

ksundong commented on backend/src/main/java/com/codesquad/coco/global/exception/ErrorCode.java:
```diff
@@ -0,0 +1,58 @@
+    /**
+     * 잘못된 input
+     **/
+    INVALID_INPUT(404, "잘못된 입력"),
```
잘못된 입력이 404일까요?

choigiseong replied:
400으로 교체했습니다!

ksundong commented on backend/src/main/java/com/codesquad/coco/global/exception/ErrorCode.java:
```diff
@@ -0,0 +1,58 @@
+     * 인증 관련 예외
+     **/
+    NOT_LOGGED_ID(401, "로그인 상태가 아닙니다"),
+    UNKNOWN_DEVICE(404, "올바른 device 접근이 아닙니다"),
+    UNAUTHORIZED_JWT(404, "jwt를 확인 하세요"),
```
여기도 404가 맞을까요?

choigiseong replied:
401로 수정했습니다!

ksundong commented on backend/src/main/java/com/codesquad/coco/global/exception/ErrorCode.java:
```diff
@@ -0,0 +1,58 @@
+        return message;
+    }
+
+    public ErrorCode plusMessage(String valueOf) {
+        this.message += valueOf;
```
파라미터명을 좀 더 제대로 바꿔주시는게 좋을 것 같아요.

choigiseong replied:
message로 수정했습니다.

ksundong commented on backend/src/main/java/com/codesquad/coco/global/exception/business/NonReservationException.java:
```diff
@@ -0,0 +1,11 @@
+
+public class NonReservationException extends BusinessException {
+
+    public NonReservationException() {
```
NonReservation은 뭔가 비문에 가까운 느낌이네요.
AlreadyReserved정도가 낫지않을까요?

ksundong commented on backend/src/main/java/com/codesquad/coco/global/exception/business/notfound/NotFoundRoomException.java:
```diff
@@ -0,0 +1,10 @@
+public class NotFoundRoomException extends NotFoundException {
+
+    public NotFoundRoomException() {
+        super(ErrorCode.USER_NOT_FOUND);
```
에러코드가 뭔가 잘못된 것 같은데요?

choigiseong replied:
수정했습니다!

ksundong commented on backend/src/main/java/com/codesquad/coco/global/exception/business/notfound/NotFoundUser.java:
```diff
@@ -0,0 +1,10 @@
+
+import com.codesquad.coco.global.exception.ErrorCode;
+
+public class NotFoundUser extends NotFoundException {
```
네이밍이 달라졌네요.

choigiseong replied:
허허.. 개판이네요..

ksundong commented on backend/src/main/java/com/codesquad/coco/image/Image.java:
```diff
@@ -0,0 +1,48 @@
+    private Long id;
+
+    private String url;
+    private String type;
```
Enum을 써보는건 어땠을까요?

choigiseong replied:
처음부터 enum으로 생각했어야 했는데, 왜 생각하지 못했을까요. 다음부터 적용하겠습니다.

ksundong commented on backend/src/main/java/com/codesquad/coco/oauth/ServerKey.java:
```diff
@@ -0,0 +1,19 @@
+
+    public ServerKey(String key) {
+        this.jwtServerKey = key;
+        logger.debug(jwtServerKey);
```
이래도 괜찮을까? 하는 생각이 들긴하네요.
객체 안에서 한 번도 로깅해볼 생각을 안해봤어서...
일단 써보시죠!

ksundong commented on backend/src/main/java/com/codesquad/coco/oauth/gitoauth/GitOauth.java:
```diff
@@ -0,0 +1,39 @@
+import static com.codesquad.coco.oauth.gitoauth.GitURI.USER_INFO_URI;
+
+@Component
+public class GitOauth extends Oauth {
```
상태때문에 추상클래스를 써주신걸까요?

ksundong replied:
뭔가 메서드가 구현되어있는 것 같긴한데, 인터페이스를 사용하는게 더 낫지 않았을까? 하는 생각이 듭니다.

choigiseong replied:
`Oauth`는 기본 class입니다.
`accessToken`을 받는 부분이나 `userInfo`를 받는 부분은 기본 동작으로 서비스마다(git, google) 같다고 생각했습니다.  그래서 `Oauth `내부에서 `accessToken`과 `userInfo`를 얻는 부분을 구현했습니다.
근데 막상 구현하다 보니 `user`정보를 얻는 부분은 서비스마다 다를 수 있겠다고 판단을 했습니다. 그래서 `Oauth`에서 정의한 `userInfo`를 상속받는 `GitUserInfo`를 만들어 받는 식으로 처리했습니다.

ksundong replied:
구현이 다르다면 그냥 인터페이스만 두고 구현상속은 하지 않는 방향을 추천드릴게요.

ksundong replied:
결국 부모클래스에 구현이 묶여버리는 설계가 나와서 추천드리지 않아요.

ksundong commented on backend/src/main/java/com/codesquad/coco/oauth/gitoauth/GitOauth.java:
```diff
@@ -0,0 +1,39 @@
+        return getTemplate().exchange(userInfoUri, HttpMethod.GET, userInfo, GitUserInfoDTO.class).getBody();
+    }
+
+    public void changeType(GitHubDeviceType gitHubDeviceType) {
```
왜 바꿀 수 있나요?

choigiseong replied:
ios와 fe는 다른 clientId가 필요하기에 변경했습니다.
다른 방법으로는 각각 인스턴스를 만들어 사용할까도 생각해 봤지만 변경하는 비용이 더 적겠다고 생각해 이렇게 했습니다.

ksundong replied:
저는 이 부분이 잘 이해가 안가요 왜 다른 clientId가 필요한가용??

choigiseong replied:
git oauth app은 하나의 redirectUri만 등록 가능한 게 문제의 원인입니다.
ios측과 fe측이 code를 받을 redirectUri가 다르기 때문입니다.


fe측만 있을 경우는 localhost:3000을 redirectUri로 해버리면 됐었는데,
ios도 같이 협업하게 되어서 localhost:3000으로 등록을 해버리면 code를 받을 수 없더라고요.
그래서 fe는 localhost:3000이 redirect uri가 되고
ios측은 앱 개별의 uri로 지정했습니다.


(이게 아니라면 redirectUri를 서버 쪽으로 해서 처리하는 방식이 있을 것으로 생각합니다.
ios와 fe로 가기 전에 중간 page를 만들어 버튼 형식으로 각각 전달하는?
이는 구현해 보지 않아서 가능한지는 잘 모르겠네요.)

그래서 생각한 점이 2개의 app을 만들어 각각 관리하자 였습니다.
이런 생각을 한 이유는 인프런 서비스가 
[안드로이드](https://user-images.githubusercontent.com/60220562/120885250-e9e1a900-c622-11eb-9274-c4f10b4a7ab6.png)
[브라우저](https://user-images.githubusercontent.com/60220562/120885256-eea65d00-c622-11eb-8ea9-d872d3b6dee3.png)
각각 다른 git oauth app을 쓰고 있기에 이것을 보고 따라하게 되었습니다.

git로그인이 가능한 서비스는 더 못찾아서 인프런만 모티브로 하였는데, 다른 서비스에서는 어떻게 풀어가는지 궁금하기도 합니다.

ksundong commented on backend/src/main/java/com/codesquad/coco/room/RoomController.java:
```diff
@@ -0,0 +1,40 @@
+
+
+    //todo 버저닝 알려주기
+    @GetMapping("/search/prices")
```
search라는 동사를 넣어주신 이유가 있나요?

choigiseong replied:
구글 검색 uri가 search를 사용하길래 따라했습니다.

ksundong replied:
사실 구글 검색의 경우에는 RESTful 하다고 표방하지 않는 만큼 참고하시면 힘들것 같네요.

ksundong commented on backend/src/main/java/com/codesquad/coco/room/RoomDAO.java:
```diff
@@ -0,0 +1,87 @@
+public class RoomDAO {
+
+    private NamedParameterJdbcTemplate template;
+    private ImageDAO imageDAO;
```
DAO를 가져다 쓰는 이유가 뭔가요?

choigiseong replied:
이 부분은 밑의 n+1 부분과 같은 이유입니다.
room을 부를 때 완벽하게 다 가지고 오려 했습니다.
한 번에 다 가지고 오면 java코드에서는 이 메소드를 통해 어느 곳에서든 알맞게 사용할 수 있겠다고 생각했습니다.
생각해보니 DAO인데 Repository를 흉내 내고 있지 않나 생각이 드네요.

ksundong replied:
Repository와는 다르게, Repository 안에 Repository같은 형태로 구성되어있는 것 같은데요~
그냥 테이블을 조인해오면 안됐을까요?

choigiseong replied:
일대다로 구현되어 있어서(방이 일, 이미지가 다) 썸네일 한 장만 가지고 오지 않는 이상 조인 시 방 정보가 중복으로 오지 않나요? 
그래서 방 가져오는 쿼리, 이미지 전부 가지고 오는 쿼리를 짰습니다.

ksundong commented on backend/src/main/java/com/codesquad/coco/room/RoomDAO.java:
```diff
@@ -0,0 +1,87 @@
+                .addValue("price_max", roomDTO.getPriceMax());
+        List<Room> rooms = template.query(FIND_ALL_ROOM_BY_SEARCH_REQUIREMENT, parameter, new RoomMapper());
+        for (Room room : rooms) {
+            //todo n+1 을 만들것인가?
+            fillRoomWithImage(room);
+            fillRoomWithReservation(room);
```
왜 N + 1을 만들어주셨는지 궁금하네요.

ksundong commented on backend/src/main/java/com/codesquad/coco/room/model/AdditionalCost.java:
```diff
@@ -0,0 +1,73 @@
+import com.codesquad.coco.utils.CalcUtil;
+import org.springframework.data.annotation.Id;
+
+public class AdditionalCost {
```
이 클래스는 하려고 했던 것이 맘에 드네요.

choigiseong replied:
감사합니다!

ksundong commented on backend/src/main/java/com/codesquad/coco/room/model/Guest.java:
```diff
@@ -0,0 +1,42 @@
+        this.baby = baby;
+    }
+
+    public static Guest guestInfo(Integer adult, Integer child, Integer baby) {
+        adult = adult != null ? adult : 0;
+        child = child != null ? child : 0;
+        baby = baby != null ? baby : 0;
```
그냥 primitive type을 받는건 어떤가요..?

choigiseong replied:
앗. 그렇겠네요. priceMin,Max와 함께 처리하다 보니 안 해도 되는 곳에 해놨네요.
수정했습니다.

ksundong commented on backend/src/main/java/com/codesquad/coco/room/model/Location.java:
```diff
@@ -0,0 +1,45 @@
+
+    private double latitude;
+    private double longitude;
+    private City city;
```
City가 같이 있는 편이 나았나요?

choigiseong replied:
음. 현재는 어디서도 사용하지는 않지만 db와 맞춰주려는 의도였습니다.  location이 city를 외래키로 가지고 있어서 객체로 구현해봤습니다.

ksundong replied:
latitude와 longitude를 같이 다니게끔 해주면 어떨까요?

ksundong replied:
그렇군요.

ksundong commented on backend/src/main/java/com/codesquad/coco/room/model/Location.java:
```diff
@@ -0,0 +1,45 @@
+    @Id
+    private Long id;
+
+    private double latitude;
+    private double longitude;
```
이 값들은 값 객체로 표현되는게 좋지 않았을까요?

choigiseong replied:
Coordinates라는 VO를 만들었습니다. 
의미 전달이 더 좋아졌습니다!

ksundong commented on backend/src/main/java/com/codesquad/coco/room/model/Money.java:
```diff
@@ -0,0 +1,38 @@
+package com.codesquad.coco.room.model;
+
+public class Money {
```
좋은 시도네요.

ksundong commented on backend/src/main/java/com/codesquad/coco/room/model/Money.java:
```diff
@@ -0,0 +1,38 @@
+public class Money {
+
+    private static final int MIN_PRICE = 0;
+    private static final int MAX_PRICE = 999_999_999;
```
```suggestion
    public static final Money MIN_PRICE = new Money(0);
    public static final Money MAX_PRICE = new Money(999_999_999);
```

이런건 어때요?

ksundong commented on backend/src/main/java/com/codesquad/coco/room/model/Money.java:
```diff
@@ -0,0 +1,38 @@
+        this.money = money;
+    }
+
+    public static Money minPrice(Integer money) {
+        money = money != null ? money : MIN_PRICE;
+        return new Money(money);
+    }
+
+    public static Money maxPrice(Integer money) {
+        money = money != null ? money : MAX_PRICE;
+        return new Money(money);
+    }
```
뭔가 이상하지 않나요? ㅎㅎ

choigiseong replied:
이 부분은 검색 조건이 null로 들어올 때 default 값을 갖게 하려 했던 부분입니다.
Integer를 사용한 이유도 null 체크를 하기 위함이었습니다.
int로 했을 때 minPrice는 0이 들어와도 0으로 처리하면 되지만 maxPrice는 0으로 들어오면 0으로 처리해주고 null이면 조건을 안 적었다는 것이니 default인 max값을 주기 위함이었는데 좀 이상한가요?...

ksundong replied:
money를 null을 안받아주면 되지 않을까요?
MYSQL에는 null일 때 지정된 값을 넣을 수 있는 Function도 있습니다.

choigiseong replied:
처음 fe와 ios 분들과 이야기 할 때 null 값도 받아 주기로 했습니다.
이 부분은 검색할 때 검색 조건을 다 입력하지 않아도 검색이 되게 하기 위함이었는데요.

처음 부터 동적 쿼리를 썼다면 이런 식으로 풀지 않았을 텐데 이 부분은 좀 아쉬웠습니다.

ksundong commented on backend/src/main/java/com/codesquad/coco/room/model/Review.java:
```diff
@@ -0,0 +1,38 @@
+    @Id
+    private Long id;
+
+    private double star;
+    private double review;
```
왜 double 형인지, Entity로 관리되어야 하는지 궁금하네요.

choigiseong replied:
리뷰는 구현하지 않기로 해서 모양만 만들었습니다.

ksundong commented on backend/src/test/java/com/codesquad/coco/ReservationTest.java:
```diff
@@ -0,0 +1,138 @@
+import static org.junit.jupiter.api.Assertions.*;
+
+
+public class ReservationTest {
```
시도 자체로 정말 높은 점수 드리고 싶네요 ㅎㅎ

ksundong commented on backend/src/test/java/com/codesquad/coco/ReservationTest.java:
```diff
@@ -0,0 +1,138 @@
+    @Test
+    @DisplayName("기본 요금 추가가 잘 되는지")
+    void totalPrice() {
+        Room room = getRoom();
```
Room에 대한 가격을 인자로 넘겨준다면 좀 더 추측가능한 테스트코드가 되지않을까요?

choigiseong replied:
그러게요.. 저도 값을 한 번 더 확인하게 되네요.

ksundong commented on backend/src/main/java/com/codesquad/coco/room/model/dto/SearchPriceDTO.java:
```diff
@@ -0,0 +1,45 @@
+    private static final LocalDate DEFAULT_CHECK_IN = LocalDate.now();
+    private static final LocalDate DEFAULT_CHECK_OUT = LocalDate.now().plusMonths(3L);
+
+    protected LocalDate checkIn;
+    protected LocalDate checkOut;
+    protected String cityName;
```
왜 Protected일까요?

ksundong commented on backend/src/main/java/com/codesquad/coco/room/model/dto/SearchPriceDTO.java:
```diff
@@ -0,0 +1,45 @@
+    protected LocalDate checkOut;
+    protected String cityName;
+
+    @ConstructorProperties({"check-in", "check-out", "city-name"})
```
저는 처음 보는 애노테이션이군요 ㅎㅎ

choigiseong replied:
rest api는 하이픈을 사용하라고 해서 사용했는데,
받을 때 바로 변환이 안되더라고요.
converter를 사용하거나 위 처럼 하라고 해서 적용 시켰습니다.

ksundong replied:
어디에서 나온 내용일까요?
케밥케이스의 사용은 처음 듣는 얘기네요.

choigiseong replied:
https://stackoverflow.com/questions/10302179/hyphen-underscore-or-camelcase-as-word-delimiter-in-uris 
https://specs.openstack.org/openstack/api-wg/guidelines/naming.html

혹시 제가 이해를 잘못한 것일까요? 구글 번역의 힘으로는 하이픈을 사용하라고 이해를 했습니다.

ksundong replied:
@ChoiGiSung 
제가 다시 리뷰를 보고 있는데 답변을 안달아서 답변 달아드립니다.
리소스에는 하이픈을 달아주는 것이 맞지만
응답 필드에서는 정해진 규칙은 없고 일반적으로 snake case를 많이씁니다. 정하기 나름이라 camel case도 많이쓰고, kebab case는 처음보는 형태이긴 했어요.

ksundong commented on backend/src/main/java/com/codesquad/coco/room/model/dto/SearchRoomDTO.java:
```diff
@@ -0,0 +1,72 @@
+import java.beans.ConstructorProperties;
+import java.time.LocalDate;
+
+public class SearchRoomDTO extends SearchPriceDTO {
```
이래서군요..
추상클래스로 바꿔주세요 ㅎㅎ

choigiseong replied:
controller에서 사용하기에 쉽지 않을 것 같습니다
Search를 최상위로 두고 다형성을 controller에 적용 시켜 보려 했는데, 타입을 명시하지 않는 한 불가능하다는 글을 봤습니다.
https://stackoverflow.com/questions/42965921/use-interface-in-spring-controller-method-argument

ksundong replied:
이런 문제가 있었군요.. 제 리뷰는 무시하셔도 괜찮을 것 같습니다~

ksundong commented on backend/src/main/java/com/codesquad/coco/room/model/dto/SearchRoomDTO.java:
```diff
@@ -0,0 +1,72 @@
+    private Money priceMin;
+    private Money priceMax;
+
+    @ConstructorProperties({"check-in", "check-out", "city-name", "adult", "child", "baby", "price-min", "price-max"})
```
딱히 필요없어보이는데 작성해주신 이유가 뭘까요?

choigiseong replied:
컨트롤러에서 다형성을 이용하지 않고 있어서 각각 만들어 줬습니다.
값이 들어오는 것을 다형성으로 처리하는 방법을 모르겠어서 놓고 있었습니다.

ksundong replied:
다형성의 문제는 아닌 것 같은데 케밥케이스여서 생기는 문제같아요.

ksundong commented on backend/src/main/java/com/codesquad/coco/user/UserService.java:
```diff
@@ -0,0 +1,68 @@
+            return;
+        }
+        userDAO.insertUser(userInfo, accessToken);
+        return;
```
마지막 return은 굳이 안해줘도 되지 않을까요?

ksundong commented on backend/src/main/java/com/codesquad/coco/user/model/Reservation.java:
```diff
@@ -0,0 +1,146 @@
+
+    public boolean reservationDateCheck(LocalDate checkIn, LocalDate checkOut) {
+        //todo : 체크 아웃 날에는 체크인이 가능하다?
+        if (LocalDateUtil.betweenCheck(this.checkIn, checkIn, checkOut) ||
+                LocalDateUtil.betweenCheck(this.checkOut, checkIn, checkOut) ||
+                this.checkIn.isEqual(checkIn) ||
+                this.checkOut.isEqual(checkOut)
+        ) {
+            return false;
+        }
+        return true;
```
조건문 자체로 결과가 될 수 있지 않을까요?

choigiseong replied:
수정했습니다! 항상 받는 피드백인데 허허..

ksundong commented on backend/src/main/java/com/codesquad/coco/user/model/WishList.java:
```diff
@@ -0,0 +1,35 @@
+
+    private Long userId;
+    private Long roomId;
+    private boolean wish;
```
boolean형이 아니라 그냥 데이터의 유무로도 알 수 있을 것 같다는 생각이 드네요 ㅎㅎ

ksundong commented on backend/src/main/java/com/codesquad/coco/utils/CalcUtil.java:
```diff
@@ -0,0 +1,12 @@
+package com.codesquad.coco.utils;
+
+public class CalcUtil {
```
개인적으로 Fullname을 적어주시는게 좋을 것 같아요.

ksundong commented on backend/src/main/java/com/codesquad/coco/utils/CalcUtil.java:
```diff
@@ -0,0 +1,12 @@
+    private CalcUtil() {
+    }
+
+    public static int percentCalc(int target, int percent) {
+        return (int) (target * (percent / 100.0));
+    }
```
잘 와닿지 않네요.
좀 더 남이 봤을 때 직관적으로 설계하는 시도를 해보면 좋을 것 같아요.

choigiseong replied:
bat도 계속 말해줬던 부분인데, 소수점으로 저장하고 바로 곱하는 게 맞을 거 같습니다.

ksundong replied:
네이밍도 좀 더 생각해보시면 좋을 것 같아요.

ksundong commented on backend/src/main/java/com/codesquad/coco/utils/DTOConverter.java:
```diff
@@ -0,0 +1,124 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class DTOConverter {
```
요건 없애는게 좋을 것 같네요.

choigiseong replied:
삭제했습니다!

ksundong commented on backend/src/main/java/com/codesquad/coco/utils/GetPropertyUtil.java:
```diff
@@ -0,0 +1,32 @@
+import org.springframework.context.annotation.PropertySource;
+
+@PropertySource("classpath:/secretKey.properties")
+public class GetPropertyUtil {
```
그냥 PropertyUtil이었어도 상관없을 것 같아요.
그리고 뭔가 빠졌죠.

ksundong replied:
만약에 ApplicationContext가 필요했다면, 굳이 Util이어야 했나 싶네요.

choigiseong replied:
private된 생성자를 추가해줬습니다. 음 util말고는 딱히 이름이 떠오르지 않아서 지었습니다.

ksundong commented on backend/src/main/java/com/codesquad/coco/utils/mapper/CityMapper.java:
```diff
@@ -0,0 +1,17 @@
+import java.sql.ResultSet;
+import java.sql.SQLException;
+
+public class CityMapper implements RowMapper<City> {
```
람다식으로도 만들어줄 수 있고 Bean으로 관리해도 됩니다.
매 번 생성해 줄 필요는 없어요.

ksundong commented on backend/src/main/java/com/codesquad/coco/utils/sql/CitySQL.kt:
```diff
@@ -0,0 +1,8 @@
+package com.codesquad.coco.utils.sql
+
+const val FIND_ALL : String = """
```
IntelliJ는 주석이나 String을 code로 인식하는 기능을 이용하면 Code Completion등의 기능을지원해줍니다.

ksundong commented on backend/src/main/java/com/codesquad/coco/utils/sql/CitySQL.kt:
```diff
@@ -0,0 +1,8 @@
+
+select c.name, c.image_url
+from city c
+
```
조건절이나 LIMIT 없이 사용은 지양해주세요.

choigiseong replied:
10개로 제한했습니다!

ksundong commented on backend/src/main/java/com/codesquad/coco/utils/sql/RoomSQL.kt:
```diff
@@ -0,0 +1,67 @@
+package com.codesquad.coco.utils.sql;
+
+const val FIND_ALL_PRICE_BY_SEARCH_REQUIREMENT: String = """
```
이 파일 같이 일하는 동료가 짰다고 생각해보시면, 개선 포인트를 찾을 수 있지 않을까요?

choigiseong replied:
몇 개를 다른 sql로 이동시켰습니다!

ksundong replied:
그것 말고도 SQL이 엄청 읽기 힘들었을 것 같아요~

ksundong commented on backend/src/test/java/com/codesquad/coco/ReservationTest.java:
```diff
@@ -0,0 +1,138 @@
+    }
+
+    @Test
+    @DisplayName("날짜가 중간에 들어가는지")
```
좀 더 상세하게 적어주면 좋을 것 같네요.

ksundong reviewed (changes requested):
안녕하세요! Coco! 리뷰어 Dion입니다.
리뷰 할 양이 많아서 조금 걸렸습니다. 양해부탁드려요 ㅠㅠ

프로젝트를 진행하느라 정말 고생많으셨습니다.
특히 단위테스트를 시도해보신 점 정말 좋은 것 같아요. 그리고 잘 하셨다고 생각됩니다.

사실 리뷰를 원하는 로직들이 무엇을 리뷰를 원하는건지 잘 이해가가지 않았습니다.
어떤 부분에 대한 리뷰를 요청하신건지 특히 Enum부분은 이해가 가지 않아서 별다른 리뷰역시 남기지 않았습니다.

고생하셨습니다~ 제가 생각하는 개선방향을 코멘트에 남겨두었으니 확인해주세요.
확인 하신 후에 생각하는 개선 포인트를 작성해주시면, 확인하고 머지해드리겠습니다~

choigiseong committed:
refactor : 총 금액 계산 리팩토링

- additionalCost에게 한 번에 계산된 금액을 받게 되었다
- 책임을 따로따로 보여주지 않고 부가비용을 받는 책임 하나만 노출시킨다

choigiseong committed:
refactor : reservation을 일급 컬렉션으로 변경

- 일급컬렉션으로 인해 의도가 명확해 졌다

choigiseong committed:
refactor : 책임을 이전

- 일급 컬렉션으로 관리해서 검증이 그쪽으로 넘어갔다
- room option도 아예 책임을 넘겨버렸다

choigiseong committed:
feat : city관련 수정 사항 변경

- null을 예방하기 위해 초기화
- final 붙여주기
- requestMapping 사용

choigiseong committed:
refactor : DTOConvertor 삭제

- 모든 변환 로직은 dto에게 위임했다

choigiseong committed:
fix : http 상태 코드 변경

- jwt가 만료된거면 로그인이 되어있지 않은 상태이다

choigiseong committed:
refactor : 각종 이름 변경

- notFoundUser는 Exception을 붙여주자
- AlreadyReserved로 변경

choigiseong committed:
refactor : 원시값으로 변경

- null이면 어차피 0 이들어온다

choigiseong committed:
feat : 좌표를 VO로 표현

- vo로 표현했습니다

choigiseong committed:
refactor : money 팩토리 메소드 변경

- 내부 구조를 변경

choigiseong committed:
feat : 좌표 VO 추가

- vo로 사용

choigiseong committed:
Merge branch 'feature_BE/46' of https://github.com/ChoiGiSung/airbnb into feature_BE/46

choigiseong committed:
refactor : if문 자체로 판단

- true,false말고 반환을 바로 한다

choigiseong committed:
refactor : PropertyUtil 이름 변경

- get을 뺐다

choigiseong committed:
feat : city select 에 조건 걸기

- 10개로 제한

choigiseong committed:
feat : sql별로 분할

- select 값에 맞게 옮겼다

choigiseong committed:
feat : displayName 변경

choigiseong committed:
Merge pull request #49 from ChoiGiSung/feature_BE/46

리뷰 사항 반영 -1

ksundong reviewed (approved):
피드백 반영하고 적용하시느라 고생하셨습니다~

choigiseong commented:
감사합니다 디온.
몇몇 부분에 확신이 없었는데 도움이 되었습니다.
client id에 관한 부분은 다시 코멘트를 달아놨습니다.
